File: go-web-basics/code-samples/handlers/hit-counter.go
package main

import (
	"fmt"
	"log"
	"net/http"
)

type Counter int

//This method makes a Counter satisfy the Handler interface
func (h *Counter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	*h++
	fmt.Fprintf(w, "This app's hit count: %d", *h)
}

//A handler function for the /sloths route
func slothsRule(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "Sloths rule!")
}

func main() {
	//Initialize our counter
	counter := Counter(0)

	mux := http.NewServeMux()
	//Use the counter as a Handler
	mux.Handle("/", &counter)

	//Use http.HandlerFunc to convert slothsRule to a Handler
	slothsRuleHandler := http.HandlerFunc(slothsRule)
	mux.Handle("/sloths", slothsRuleHandler)

	//Logs what URL the request is to and then sends the request to mux
	//by calling its ServeHTTP method.
	logAndServe := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Println("Request to " + r.URL.String())
		mux.ServeHTTP(w, r)
	})

	server := &http.Server{
		Addr:    ":1123",
		Handler: logAndServe,
	}
	server.ListenAndServe()
}

File: go-web-basics/code-samples/handlers/hit-counter.js
var http    = require('http'),
    express = require('express')

var app = express()
var counter = 0

//Logger middleware, run by all requests
app.use(function(req, res, next){
  console.log('Request to', req.url)
  next()
})

//Serve the /sloths route
app.use('/sloths', function(req, res){
  res.send('Sloths rule!')
})

//Catch-all route, serves the hit counter
app.use(function(req, res) {
  counter++
  res.send("This app's hit count: " + counter)
})

http.createServer(app).listen(1123)
